src/WorldIDIdentityManagerImplV1.sol

```diff
@@ -362,6 +513,40 @@ contract WorldIDIdentityManagerImplV1 is OwnableUpgradeable, UUPSUpgradeable, IW
         hash = keccak256(bytesToHash);
     }
 
+    /// @notice Calculates the input hash for the identity update verifier.
+    /// @dev Implements the computation described below.
+    ///
+    /// @param preRoot The root value of the tree before the updates were made.
+    /// @param postRoot The root value of the tree after the udpates were made.
+    /// @param identities The identity structures providing the identity data.
+    ///
+    /// @return hash The input hash calculated as described below.
+    ///
+    /// We keccak hash all input to save verification gas. The inputs are arranged as follows:
+    ///
+    /// preRoot || postRoot || id[0].leafIndex || id[0].oldCommitment || id[0].newCommitment || ... || id[n].leafIndex || id[n].oldCommitment || id[n].newCommitment
+    ///   256   ||    256   ||        32       ||         256         ||          256        || ... ||        32       ||         256         ||          256
+    ///
+    /// where:
+    /// - `id[n] == identities[n]`
+    /// - `n == batchSize - 1`
+    function calculateIdentityUpdateInputHash(
+        uint256 preRoot,
+        uint256 postRoot,
+        Identity[] calldata identities
+    ) public view virtual onlyProxy onlyInitialized returns (bytes32 hash) {
+        bytes memory identityBytes = new bytes(0);
```

Comment:
it would probably be faster to pre-alloc this, given we know the exact size needed?

Reply:
I think it would actually be a touch more expensive. I can't assign from `abi.encode` into a slice, so I have to allocate an additional buffer _anyway_ and then copy.